sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -173,7 +203,36 @@ class GpuSortAggregateMeta(
         case "partial" => if (hashAggMode.contains(Final) || hashAggMode.contains(Complete)) {
           // replacing only Partial hash aggregates, so a Final or Commplete one should not replace
           willNotWorkOnGpu("Replacing Final or Complete hash aggregates disabled")
-         }
+        }
+          // In partial mode, if there are non-distinct functions and multiple distinct functions,
+          // non-distinct functions are computed using the First operator. The final result would be
+          // incorrect for non distinct functions for partition size > 1. Reason for this is - if
+          // the first batch computed and sent to CPU doesn't contain all the rows required to
+          // compute non distinct function(s), then Spark would consider that value as final result
+          // (due to First)Falling back to CPU for this special case.
+          if (agg.aggregateExpressions.exists(e => e.aggregateFunction.isInstanceOf[First])) {
+            agg.aggregateExpressions.foreach(e => {
+              // Check if there is an `If` within `First`. This is included in the plan only when
+              // multiple distinct is present in the query. We fall back to CPU in this case when
+              // the references of `If` is an aggregate. We cannot call `isDistinct` here on
+              // aggregateExpressions to get the total number of distinct functions. If there are
+              // multiple distinct the plan is rewritten by `RewriteDistinctAggregates` where
+              // regular aggregations and every distinct aggregation is calculated in a separate
+              // group.
+              if (e.aggregateFunction.isInstanceOf[First] &&
+                e.aggregateFunction.asInstanceOf[First].child.isInstanceOf[If]) {
+                // Get the aggregate from references of `If`
+                val aggRefs = e.aggregateFunction.references
+                val isMultipleDistinctAggs = validateAggregate(aggRefs)
+
+                if (isMultipleDistinctAggs) {
+                  willNotWorkOnGpu("Aggregate of non-distinct functions with multiple distinct " +
+                    "functions is non deterministic for non-distinct functions as it is " +
+                    "computed using First.")
+                }
+              }
+            })
+          }
```

Comment:
You can make it even simpler, because we have a noop on things that don't match.

```
agg.aggregateExpressions.map(e => e.aggregateFunction).exists { func =>
  func match {
    case First(If, _) if (validateAggregate(func.references)) =>
      ...
    case _ =>
  }
}

```